### PR TITLE
Add an empty standalone model for testing purposes

### PR DESCRIPTION
**Pull request by mikapfl**
## :bird: Description of this PR :bird:

Add an empty standalone test model. The idea is that start functions are run as normal, the model is compiled and configured as usual, but then the model doesn't actually run. Instead, the latest automated model test (AMT) result is simply copied to `fulldata.gdx` and the model finishes. Afterwards, reporting is run as usual / as configured.

The reason why I need this is to write somewhat performant tests for the coupled REMIND-MAgPIE system, where we have to run REMIND and MAgPIE for multiple iterations each, so that we don't have the time to actually run the models. However, I could imagine that it is also useful for MAgPIE standalone tests, e.g. to test changes in start scripts or reporting functions.

## :wrench: Checklist for PR creator :wrench:

- [x] Label pull request [from the label list](https://github.com/magpiemodel/magpie/labels).
  - **Low risk**: Simple bugfixes (missing files, updated documentation, typos) or changes  in start or output scripts
  - **Medium risk**: Uncritical changes in the model core (e.g. moderate modifications in non-default realizations)
  - **High risk**: Critical changes in model core or default settings (e.g. changing a model default or adjusting a core mechanic in the model)

- [x] Self-review own code
  - No hard coded numbers and cluster/country/region names.
  - The new code doesn't contain declared but unused parameters or variables.
  - [`magpie4`](https://github.com/pik-piam/magpie4) R library has been updated accordingly and backwards compatible where necessary.
  - `scenario_config.csv` has been updated accordingly (important if `default.cfg` has been updated)

- [x] Document changes 
  - Add changes to `CHANGELOG.md`
  - Where relevant, put In-code documentation comments
  - Properly address updates in interfaces in the module documentations
  - run [`goxygen::goxygen()`](https://github.com/pik-piam/goxygen) and verify the modified code is properly documented

- [x] Perform test runs
  - **Low risk**: 
    - Run a compilation check via `Rscript start.R --> "compilation check"`
  - **Medium risk**: 
    - Run test runs via `Rscript start.R --> "test runs"`
    - Check logs for errors/warnings
  - **High risk**:
    - Run test runs via `Rscript start.R --> "test runs"`
    - Check logs for errors/warnings
    - Default run from the PR target branch for comparison
    - Provide relevant comparison plots (land-use, emissions, food prices, land-use intensity,...)

### :chart_with_downwards_trend: Performance changes :chart_with_upwards_trend:
  
  - This PR's `Rscript start.R runscripts=extra/empty_model.R submit=direct` :  8.3 mins

## :rotating_light: Checklist for reviewer :rotating_light:

- PR is labeled correctly
- Code changes look reasonable
  - No hard coded numbers and cluster/country/region names.
  - No unnecessary increase in module interfaces
  - model behavior/performance is satisfactory.
- Changes are properly documented
  - `CHANGELOG` is updated correctly
  - Updates in interfaces have been properly addressed in the module documentations
  - In-code documentation looks appropriate
- [x] content review done (at least 1)
- [ ] RSE review done (at least 1)

### CODE REVIEW

**Comment by mikapfl** [comment]
We still need a couple of iterations until this can really be used in the REMIND-MAgPIE tests, but it would be nice to get an early review if the direction looks good.

**Comment by tscheypidi** [comment]
Overall direction looks good to me. Only problem I see is that the `empty_test_model.gms` probably will get quickly out of sync with `main.gms`. To avoid it one could think about letting `empty_model.R` construct this file by reading in `main.gms`, performing the required modifications and then storing it under the new name. What do you think?

**Comment by mikapfl** [comment]
> To avoid it one could think about letting empty_model.R construct this file by reading in main.gms, performing the required modifications and then storing it under the new name. What do you think?

Possible, sure. Probably takes a bit longer until this code then needs changes because it is out of sync with `main.gms`, but when you have to change it, it will be more complex and harder to understand than the copying necessary for updating `empty_test_model.gms`. I'm not sure if this is a better trade-off, but I can implement it either way.

These synchronization problems don't arise if the necessary code is (behind an `if`) in `main.gms`, which is why we took that route in REMIND.

**Review by tscheypidi — APPROVED** [review]
looks good

**Comment by mikapfl** [comment]
Who else should I ask to get to two approving reviews?

**Comment by tscheypidi** [comment]
normally another non-RSE, but in this case I think @pfuehrlich-pik makes most sense. I add him as reviewer

File: scripts/start/extra/empty_model.R [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# |  (C) 2023 Potsdam Institute for Climate Impact Research (PIK)
```
Comment (pfuehrlich-pik) [inline scripts/start/extra/empty_model.R:1]: Not sure why, but our header usually has the 2008 in it. Is this actually disadvantageous from a copyright law perspective?
```suggestion
# |  (C) 2008 - 2023 Potsdam Institute for Climate Impact Research (PIK)
```

File: scripts/start_functions.R [hunk 2]
```diff
@@ -589,3 +589,35 @@ getReportData <- function(path_to_report_bioenergy, mute_ghgprices_until = "y201
     .emissionPrices(ghgmag, mute_ghgprices_until)
   }
 }
+
+# Will not actually solve the model: after compilation, this just copies the results
+# of a previous run, useful for testing compilation and input/output handling.
+# Used in scripts/start/extra/empty_model.R and tests for REMIND-MAgPIE coupling.
+configureEmptyModel <- function(cfg, inputGdxPath) {
+    message("Configuring to use empty MAgPIE model, reproduces prior run ", inputGdxPath)
+    originalModel <- file(cfg$model, "r")
```
Comment (pfuehrlich-pik) [inline scripts/start_functions.R:598]:
```suggestion
    originalModel <- withr::local_connection(file(cfg$model, "r"))
```
This ensures the file handle is properly closed when the function exits.

File: scripts/start_functions.R [hunk 3]
```diff
@@ -589,3 +589,35 @@ getReportData <- function(path_to_report_bioenergy, mute_ghgprices_until = "y201
...
+    emptyModelFile <- "standalone/empty_test_model.gms"
+    emptyModel <- file(emptyModelFile, "w")
```
Comment (pfuehrlich-pik) [inline scripts/start_functions.R:600]:
```suggestion
    emptyModel <- withr::local_connection(file(emptyModelFile, "w"))
```
This ensures the file handle is properly closed when the function exits.

File: scripts/start/extra/empty_model.R (hunk 1 above)
Comment (mikapfl) [reply scripts/start/extra/empty_model.R:1]: It is a lie. I don't like to lie.

**Review by pfuehrlich-pik — APPROVED** [review]

**Pull request merged**